### PR TITLE
Migrate to AndroidX

### PR DESCRIPTION
**Pull request by samtstern**

### CODE REVIEW

**Comment by samtstern** [comment]
@SUPERCILEX do you have any experience converting a library to AndroidX?  This is my first go at it and I am sure it's a minefield.

File: gradle.properties [hunk 1]
```diff
@@ -21,3 +21,5 @@ org.gradle.configureondemand=true
 
 # Required by Robolectric 4.x
 android.enableUnitTestBinaryResources=true
+android.useAndroidX=true
```
Comment (SUPERCILEX) [inline gradle.properties]: I would probably add a space before this property so it doesn't look like it's required for robolectric.

**Review by SUPERCILEX — APPROVED** [review]
Nah, this looks great! 😊 The alphas were bad because of weird bugs, but all you have to do today is update the imports. I just checked my migration commits and they all deal with adding Kotlin extensions everywhere... keeping up in Java land is quite a bit easier these days. 😅

**Comment by samtstern** [comment]
@henriquenfaria what do you think?

Also this would be version `3.0.0`, is there anything else we should break while we have the chance?

File: easypermissions/src/test/java/pub/devrel/easypermissions/AppSettingsDialogTest.java [hunk 2]
```diff
@@ -187,15 +187,14 @@ public void shouldNegativeListener_whenClickingPositiveButtonFromSupportFragment
     private void setUpActivityAndFragment() {
         activityController = Robolectric.buildActivity(TestActivity.class)
                 .create().start().resume();
-        fragmentController = SupportFragmentController.of(new TestFragment())
-                .create().start().resume();
+        fragmentController = new FragmentController<>(TestFragment.class);
 
         spyActivity = Mockito.spy(activityController.get());
-        spyFragment = Mockito.spy(fragmentController.get());
+        spyFragment = Mockito.spy(fragmentController.resume());
```
Comment (henriquenfaria) [inline easypermissions/src/test/java/pub/devrel/easypermissions/AppSettingsDialogTest.java]: Can we do the same for the activityController? If so, we will be able to get rid of the tearDownActivityAndFragment method.

File: easypermissions/build.gradle [hunk 3]
```diff
@@ -33,14 +33,16 @@ junitJacoco {
...
+    api "androidx.core:core:1.0.1"
+    api "androidx.fragment:fragment:1.0.0"
 
     testImplementation 'junit:junit:4.12'
     testImplementation 'com.google.truth:truth:0.42'
     testImplementation 'org.robolectric:robolectric:4.1'
-    testImplementation 'org.robolectric:shadows-supportv4:4.1'
+    // TODO
+//    testImplementation 'org.robolectric:shadows-supportv4:4.1'
+    testImplementation 'androidx.fragment:fragment-testing:1.1.0-alpha03'
```
Comment (henriquenfaria) [inline easypermissions/build.gradle]: Maybe use the previous stable version instead of the alpha one?

File: easypermissions/build.gradle [hunk 4]
```diff
@@ -33,14 +33,16 @@ junitJacoco {
...
+    api "androidx.appcompat:appcompat:1.0.2"
+    api "androidx.core:core:1.0.1"
+    api "androidx.fragment:fragment:1.0.0"
 
     testImplementation 'junit:junit:4.12'
     testImplementation 'com.google.truth:truth:0.42'
     testImplementation 'org.robolectric:robolectric:4.1'
-    testImplementation 'org.robolectric:shadows-supportv4:4.1'
+    // TODO
+//    testImplementation 'org.robolectric:shadows-supportv4:4.1'
```
Comment (henriquenfaria) [inline easypermissions/build.gradle]: Why you had to remove it? Was something wrong with the SupportFragmentController after the migration?

**Review by henriquenfaria — APPROVED** [review]
Looks nice! The jetfier should migrate our transitive dependencies to androidx as well, I think we are safe.
Just added a few comments.

File: easypermissions/build.gradle (hunk 4 above)
Comment (samtstern) [reply easypermissions/build.gradle]: This library is not compatible with AndroidX and Jetifier (a lot of people online report the same problem).  So I upgraded to the new test support library and then made a small wrapper to mimic the old behavior.

File: easypermissions/build.gradle (hunk 3 above)
Comment (samtstern) [reply easypermissions/build.gradle]: I actually don't see any stable versions ever released on maven.google.com ... seems like `1.1.0-alpha01` was the first.  However I don't mind taking an `alpha` dependency in test only, although I agree with you that stable would be better.

File: gradle.properties (hunk 1 above)
Comment (samtstern) [reply gradle.properties]: Good call.

**Comment by samtstern** [comment]
Just for clarity, here's what the 3.0 pom would look like:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>pub.devrel</groupId>
  <artifactId>easypermissions</artifactId>
  <version>3.0.0-SNAPSHOT</version>
  <packaging>aar</packaging>
  <name>EasyPermissions</name>
  <description>A wrapper library for basic Android M system permissions logic</description>
  <url>https://github.com/googlesamples/easypermissions</url>
  <inceptionYear>2015</inceptionYear>
  <licenses>
    <license>
      <name>The Apache Software License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
      <distribution>repo</distribution>
    </license>
  </licenses>
  <developers>
    <developer>
      <name>Google</name>
    </developer>
  </developers>
  <scm>
    <connection>https://github.com/googlesamples/easypermissions.git</connection>
    <url>https://github.com/googlesamples/easypermissions</url>
  </scm>
  <dependencies>
    <dependency>
      <groupId>androidx.appcompat</groupId>
      <artifactId>appcompat</artifactId>
      <version>1.0.2</version>
    </dependency>
    <dependency>
      <groupId>androidx.core</groupId>
      <artifactId>core</artifactId>
      <version>1.0.1</version>
    </dependency>
    <dependency>
      <groupId>androidx.fragment</groupId>
      <artifactId>fragment</artifactId>
      <version>1.0.0</version>
    </dependency>
  </dependencies>
</project>
```

**Comment by nimamoradi** [comment]
I ran to problem to this error in andriodx, from 3.0 pom it issus from this commit.
<pre>
[ERROR: Failed to resolve: fragment
Affected Modules: app


ERROR: Failed to resolve: core
Affected Modules: app](url)
</pre>
my gradle file is this:
<pre>
implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'androidx.core:core-ktx:1.1.0'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test:runner:1.2.0'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'

    implementation 'io.reactivex.rxjava2:rxjava:2.0.5'
    implementation 'io.reactivex.rxjava2:rxandroid:2.0.1'

    implementation('io.socket:socket.io-client:1.0.0') {
        exclude group: 'org.json', module: 'json'
    }
    implementation "com.google.android.material:material:1.0.0"
    implementation "androidx.legacy:legacy-support-v4:1.0.0"
    implementation 'org.webrtc:google-webrtc:1.0.28513'

    implementation 'com.myhexaville:smart-image-picker:1.0.4'
</pre>

and i have added google() dependency in the project gradle

**Review by adlawan — COMMENTED** [review]
Great job

**Review by privacyquestions — APPROVED** [review]

**Comment by Duresa6** [comment]
@

**Review by Duresa6 — APPROVED** [review]

**Pull request merged**